drivers/sensor/wsen_itds_2533020201601/wsen_itds_2533020201601_trigger.c

```diff
@@ -0,0 +1,562 @@
+/*
+ * Copyright (c) 2023 Würth Elektronik eiSos GmbH & Co. KG
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT we_wsen_itds_2533020201601
+
+#include <zephyr/logging/log.h>
+
+#include "wsen_itds_2533020201601.h"
+
+LOG_MODULE_DECLARE(WSEN_ITDS_2533020201601, CONFIG_SENSOR_LOG_LEVEL);
+
+/* Enable/disable interrupt handling for tap, free-fall, delta/wake-up */
+static inline int itds_2533020201601_setup_interrupt(const struct device *dev,
+						     const struct gpio_dt_spec *pin, bool enable)
+{
+	unsigned int flags = enable ? GPIO_INT_EDGE_TO_ACTIVE : GPIO_INT_DISABLE;
+
+	return gpio_pin_interrupt_configure_dt(pin, flags);
+}
+
+static inline void itds_2533020201601_handle_interrupt_1(const struct device *dev)
+{
+	struct itds_2533020201601_data *data = dev->data;
+	const struct itds_2533020201601_config *cfg = dev->config;
+
+	/* Disable interrupt handling until the interrupt has been processed */
+	itds_2533020201601_setup_interrupt(data->dev, &cfg->drdy_interrupt_gpio, false);
+
+#if defined(CONFIG_WSEN_ITDS_2533020201601_TRIGGER_OWN_THREAD)
+	k_sem_give(&data->drdy_sem);
+#elif defined(CONFIG_WSEN_ITDS_2533020201601_TRIGGER_GLOBAL_THREAD)
+	k_work_submit(&data->drdy_work);
+#endif
+}
+
+#ifdef CONFIG_WSEN_ITDS_2533020201601_EVENTS
+static inline void itds_2533020201601_handle_interrupt_0(const struct device *dev)
+{
+	struct itds_2533020201601_data *data = dev->data;
+	const struct itds_2533020201601_config *cfg = dev->config;
+
+	/* Disable interrupt handling until the interrupt has been processed */
+	itds_2533020201601_setup_interrupt(data->dev, &cfg->events_interrupt_gpio, false);
+
+#if defined(CONFIG_WSEN_ITDS_2533020201601_TRIGGER_OWN_THREAD)
+	k_sem_give(&data->events_sem);
+#elif defined(CONFIG_WSEN_ITDS_2533020201601_TRIGGER_GLOBAL_THREAD)
+	k_work_submit(&data->events_work);
+#endif
+}
+
+/* Asynchronous handling of interrupt triggered in itds_2533020201601_gpio_callback() */
+static void itds_2533020201601_process_interrupt_0(const struct device *dev)
+{
+	struct itds_2533020201601_data *data = dev->data;
+	const struct itds_2533020201601_config *cfg = dev->config;
+	ITDS_status_t itds_2533020201601_status;
+
+	/* Read status register to check for data-ready - if not using soft trigger, the
+	 * status register is also used to find out which interrupt occurred.
+	 */
+	if (ITDS_getStatusRegister(&data->sensor_interface, &itds_2533020201601_status) !=
+	    WE_SUCCESS) {
+		LOG_ERR("Failed to read status register");
+		return;
+	}
+
+#ifdef CONFIG_WSEN_ITDS_2533020201601_TAP
+	struct sensor_trigger single_tap_trigger = {.chan = SENSOR_CHAN_ALL,
+						    .type = SENSOR_TRIG_TAP};
```

Comment:
Don't create new trigger structs on the stack; store a pointer to the user's sensor trigger and pass it back